### PR TITLE
Factor out common code for handling callbacks into wrapper functions

### PR DESCRIPTION
**Pull request by RedBeard0531**
## What, How & Why?
This gets us closer to the ideal of each C++ function only expressing its core business logic without excess noise.

## ☑️ ToDos

Since this is just an internal refactor, I don't think any of the default ToDos really apply. Let me know if you disagree.

* ~[ ] 📝 Changelog entry~
* ~[ ] 📝 `Compatibility` label is updated or copied from previous entry~
* ~[ ] 🚦 Tests~
* ~[ ] 📝 Public documentation PR created or is not necessary~
* ~[ ] 💥 `Breaking` label has been applied or is not necessary~

*If this PR adds or changes public API's:*
* ~[ ] typescript definitions file is updated~
* ~[ ] jsdoc files updated~
* ~[ ] Chrome debug API is updated if API is available on React Native~

### CODE REVIEW

File: src/js_types.hpp [hunk 1]
```diff
@@ -221,6 +223,24 @@ struct Function {
...
     }
+
+    /**
+     * These wrap a js callback into a C++ callback that takes an Optional<AppError> and possibly a result.
+     * The versions that accept a result use the converter argument to convert the C++ result into a js result.
+     * The returned callbacks have the following signatures:
+     *
+     * wrap_void_callback         - void(const util::Optional<app::AppError>& error)
+     * wrap_callback_error_first  - void(const util::Optional<app::AppError>& error, auto&& result)
+     * wrap_callback_result_first - void(auto&& result, const util::Optional<app::AppError>& error)
```
Comment (kneth) [inline src/js_types.hpp:234]: Why don't we fix it in object store?

Comment (RedBeard0531) [reply src/js_types.hpp:234]: I didn't want to break compile for everyone in order to make the callback argument order consistent. At least not without discussing it with the other sdk teams. I'll ping the slack channel and see what everyone thinks. Do you mind going with this for now, and if we fix obj store, we can eliminate one of the `wrap_callback_foo_first` functions in a later commit?

**Comment by blagoev** [comment]
@tgoyne Could you share your view on this. Why not use a single instance of the EventLoopDispatcher in all places. Note that this PR is stepping on this change https://github.com/realm/realm-js/pull/3031.

In short we made a single instance of the EventLoopDispatcher to be used for all network requests in #3031 . Then this PR cleans up a lot of code using the fact that all callbacks will be invoked correctly on the main thread.

The question is why not just have a global EventLoopDispatcher somewhere and use that in similar fashion of this PR. Do you see any issues with that?

thanks

**Comment by tgoyne** [comment]
EventLoopDispatcher's design probably doesn't really make much sense any more. It was originally written as a simple adapter to wrap some pre-existing frameworks without having to redesign anything. Once you're actually reworking the code using it, there's probably a better abstraction that could be found.

**Comment by blagoev** [comment]
thanks. I think we could do that later. Not to expand the churn and the meaning of this PR.

**Pull request merged**